File: sequential_inference_pinned.py
#!/usr/bin/env python3
"""
Sequentially execute split ResNet-50 TensorRT engines with profiling.
Now uses CUDA pinned (page-locked) host memory to reduce CPU↔GPU transfer overhead.
"""

import tensorrt as trt
import pycuda.driver as cuda
import pycuda.autoinit
import numpy as np
import time
from torch.cuda import nvtx
import glob
import os

# ============================================================
# 1. Setup TensorRT runtime and load all chunk engines
# ============================================================

TRT_LOGGER = trt.Logger(trt.Logger.WARNING)
runtime = trt.Runtime(TRT_LOGGER)

ENGINE_DIR = "engines"
engine_paths = sorted(glob.glob(os.path.join(ENGINE_DIR, "*.engine")))

engines, contexts = [], []
for path in engine_paths:
    with open(path, "rb") as f:
        engine = runtime.deserialize_cuda_engine(f.read())
        ctx = engine.create_execution_context()
        engines.append(engine)
        contexts.append(ctx)
    print(f"[INFO] Loaded engine: {os.path.basename(path)}")

print(f"✅ Loaded {len(engines)} chunk engines.")

# ============================================================
# 2. Allocate reusable GPU buffers and pinned host buffers
# ============================================================

input_data = np.random.randn(1, 3, 224, 224).astype(np.float32)

max_bytes = 0
for engine in engines:
    for b in range(engine.num_bindings):
        shape = engine.get_binding_shape(b)
        if all(dim > 0 for dim in shape):
            size_bytes = int(np.prod(shape)) * 4
            max_bytes = max(max_bytes, size_bytes)

print(f"[INFO] Allocating {max_bytes / (1024**2):.2f} MB for reusable buffers...")

# GPU buffers
d_input = cuda.mem_alloc(int(max_bytes))
d_output = cuda.mem_alloc(int(max_bytes))
stream = cuda.Stream()

# Pinned host buffers per chunk (for input/output reuse)
pinned_inputs = []
pinned_outputs = []

for ctx in contexts:
    out_shape = tuple(ctx.get_binding_shape(1))
    h_in = cuda.pagelocked_empty(input_data.size, np.float32).reshape(input_data.shape)
    h_out = cuda.pagelocked_empty(int(np.prod(out_shape)), np.float32).reshape(out_shape)
    pinned_inputs.append(h_in)
    pinned_outputs.append(h_out)

# ============================================================
# 3. Helper to run one chunk
# ============================================================

def run_chunk(context, h_input, h_output, d_input, d_output, stream):
    cuda.memcpy_htod_async(d_input, h_input, stream)
    start = cuda.Event()
    end = cuda.Event()
    start.record(stream)
    context.execute_v2([int(d_input), int(d_output)])
    end.record(stream)
    cuda.memcpy_dtoh_async(h_output, d_output, stream)
    stream.synchronize()
    return start.time_till(end)  # ms

# ============================================================
# 4. Sequential inference with warmup and timing
# ============================================================

warmups = 5
num_runs = 20
all_run_times = []

print(f"\n[INFO] Starting sequential inference ({warmups} warmups, {num_runs} runs)...\n")

# Warmups
for w in range(warmups):
    nvtx.range_push(f"Warmup {w+1}")
    np.copyto(pinned_inputs[0], input_data)
    for i, ctx in enumerate(contexts):
        h_input = pinned_inputs[i] if i == 0 else pinned_outputs[i-1]
        h_output = pinned_outputs[i]
        _ = run_chunk(ctx, h_input, h_output, d_input, d_output, stream)
    nvtx.range_pop()

# Timed runs
for r in range(num_runs):
    nvtx.range_push(f"Inference {r+1}")
    np.copyto(pinned_inputs[0], input_data)
    chunk_times = []
    t0 = time.time()

    for i, ctx in enumerate(contexts):
        nvtx.range_push(f"Chunk {i+1}")
        h_input = pinned_inputs[i] if i == 0 else pinned_outputs[i-1]
        h_output = pinned_outputs[i]
        gpu_ms = run_chunk(ctx, h_input, h_output, d_input, d_output, stream)
        chunk_times.append(gpu_ms)
        nvtx.range_pop()

    t1 = time.time()
    total_gpu_time = sum(chunk_times)
    total_cpu_time = (t1 - t0) * 1000
    all_run_times.append((chunk_times, total_gpu_time, total_cpu_time))

    print(f"Inference {r+1}: total GPU {total_gpu_time:.3f} ms | total CPU {total_cpu_time:.3f} ms")

    nvtx.range_pop()

# ============================================================
# 5. Stats
# ============================================================

gpu_totals = [t[1] for t in all_run_times]
cpu_totals = [t[2] for t in all_run_times]

def summarize(name, arr):
    arr = np.array(arr)
    return f"{np.mean(arr):.3f} ± {np.std(arr):.3f} (min={np.min(arr):.3f}, max={np.max(arr):.3f})"

print("\n✅ Sequential inference complete.\n")
print(f"GPU total time (ms): {summarize('GPU', gpu_totals)}")
print(f"CPU total time (ms): {summarize('CPU', cpu_totals)}")

# ============================================================
# 6. CSV export
# ============================================================

with open("chunk_timing.csv", "w") as f:
    f.write("run_id," + ",".join([f"chunk{i+1}_gpu_ms" for i in range(len(engines))]) + ",total_gpu_ms,total_cpu_ms\n")
    for r, (chunk_times, gpu_total, cpu_total) in enumerate(all_run_times, 1):
        chunk_str = ",".join([f"{x:.4f}" for x in chunk_times])
        f.write(f"{r},{chunk_str},{gpu_total:.4f},{cpu_total:.4f}\n")

print("📝 Saved detailed timing to chunk_timing.csv")

File: sequential_inference_pinned.cpp
/*


nvcc -std=c++17 -O2 sequential_inference_gpu.cpp -o run_trt_pinned \
  -I/usr/include/x86_64-linux-gnu \
  -I/usr/local/cuda/include \
  -L/usr/lib/x86_64-linux-gnu \
  -L/usr/local/cuda/lib64 \
  -lnvinfer -lnvonnxparser -lcudart -lnvToolsExt


*/


#include <NvInfer.h>
#include <cuda_runtime.h>
#include <nvtx3/nvToolsExt.h>

#include <algorithm>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <memory>
#include <numeric>
#include <random>
#include <string>
#include <vector>

namespace fs = std::filesystem;

#define CUDA_CHECK(x)                                                   \
    do {                                                                \
        cudaError_t err = (x);                                          \
        if (err != cudaSuccess) {                                       \
            std::cerr << "CUDA error: " << cudaGetErrorString(err)      \
                      << " at " << __FILE__ << ":" << __LINE__ << "\n"; \
            std::exit(EXIT_FAILURE);                                    \
        }                                                               \
    } while (0)

struct Logger : public nvinfer1::ILogger {
    void log(Severity s, const char* msg) noexcept override {
        if (s <= Severity::kWARNING)
            std::cerr << "[TRT] " << msg << "\n";
    }
};

static size_t volume(const nvinfer1::Dims& d) {
    size_t v = 1;
    for (int i = 0; i < d.nbDims; ++i)
        v *= static_cast<size_t>(d.d[i]);
    return v;
}

static std::vector<char> loadFile(const fs::path& p) {
    std::ifstream ifs(p, std::ios::binary | std::ios::ate);
    if (!ifs) throw std::runtime_error("Failed to open: " + p.string());
    std::streamsize size = ifs.tellg();
    ifs.seekg(0, std::ios::beg);
    std::vector<char> buf(size);
    if (!ifs.read(buf.data(), size))
        throw std::runtime_error("Failed to read: " + p.string());
    return buf;
}

int main() {
    try {
        Logger logger;
        auto runtime = std::unique_ptr<nvinfer1::IRuntime>(
            nvinfer1::createInferRuntime(logger));

        // ============================================================
        // 1. Load all .engine chunks
        // ============================================================
        std::vector<fs::path> enginePaths;
        for (auto& e : fs::directory_iterator("engines"))
            if (e.path().extension() == ".engine")
                enginePaths.push_back(e.path());
        std::sort(enginePaths.begin(), enginePaths.end());
        if (enginePaths.empty())
            throw std::runtime_error("No .engine files found in ./engines");

        std::vector<std::unique_ptr<nvinfer1::ICudaEngine>> engines;
        std::vector<std::unique_ptr<nvinfer1::IExecutionContext>> contexts;

        for (auto& p : enginePaths) {
            auto bytes = loadFile(p);
            auto* e = runtime->deserializeCudaEngine(bytes.data(), bytes.size());
            if (!e) throw std::runtime_error("Failed to deserialize: " + p.string());
            engines.emplace_back(e);
            contexts.emplace_back(e->createExecutionContext());
            std::cout << "[INFO] Loaded engine: " << p.filename().string() << "\n";
        }
        int numChunks = engines.size();
        std::cout << "✅ Loaded " << numChunks << " chunk engines.\n";

        // ============================================================
        // 2. Allocate pinned host + device buffers
        // ============================================================
        const int B = 1, C = 3, H = 224, W = 224;
        nvinfer1::Dims inputDims{4, {B, C, H, W}};

        // host input initialization
        std::vector<float> tmp(volume(inputDims));
        std::mt19937 rng(42);
        std::normal_distribution<float> dist(0.f, 1.f);
        for (auto& v : tmp) v = dist(rng);

        // allocate pinned memory for inputs/outputs
        std::vector<void*> hIn(numChunks), hOut(numChunks);
        std::vector<nvinfer1::Dims> inDims(numChunks), outDims(numChunks);
        std::vector<void*> dAct(numChunks);
        void* dIn0 = nullptr;

        // create CUDA stream
        cudaStream_t stream;
        CUDA_CHECK(cudaStreamCreate(&stream));

        for (int i = 0; i < numChunks; ++i) {
            auto* e = engines[i].get();
            const char* inName = e->getIOTensorName(0);
            const char* outName = e->getIOTensorName(1);

            inDims[i] = e->getTensorShape(inName);
            outDims[i] = e->getTensorShape(outName);
            if (std::any_of(inDims[i].d, inDims[i].d + inDims[i].nbDims,
                            [](int d) { return d < 0; }))
                inDims[i] = inputDims;  // set static input
            contexts[i]->setInputShape(inName, inDims[i]);

            CUDA_CHECK(cudaMalloc(&dAct[i], volume(outDims[i]) * sizeof(float)));
            CUDA_CHECK(cudaHostAlloc(&hIn[i],
                                     volume(inDims[i]) * sizeof(float),
                                     cudaHostAllocDefault));
            CUDA_CHECK(cudaHostAlloc(&hOut[i],
                                     volume(outDims[i]) * sizeof(float),
                                     cudaHostAllocDefault));
        }
        CUDA_CHECK(cudaMalloc(&dIn0, volume(inDims[0]) * sizeof(float)));

        std::memcpy(hIn[0], tmp.data(),
                    volume(inDims[0]) * sizeof(float));

        std::cout << "[INFO] Allocated pinned host + device buffers.\n";

        // ============================================================
        // 3. Helper lambda for one chunk
        // ============================================================
        auto runChunk = [&](int i, void* dSrc, void* dDst) {
            cudaEvent_t start, stop;
            CUDA_CHECK(cudaEventCreate(&start));
            CUDA_CHECK(cudaEventCreate(&stop));
            const char* inName = engines[i]->getIOTensorName(0);
            const char* outName = engines[i]->getIOTensorName(1);
            contexts[i]->setTensorAddress(inName, dSrc);
            contexts[i]->setTensorAddress(outName, dDst);
            CUDA_CHECK(cudaEventRecord(start, stream));
            contexts[i]->enqueueV3(stream);
            CUDA_CHECK(cudaEventRecord(stop, stream));
            CUDA_CHECK(cudaEventSynchronize(stop));
            float ms = 0;
            CUDA_CHECK(cudaEventElapsedTime(&ms, start, stop));
            CUDA_CHECK(cudaEventDestroy(start));
            CUDA_CHECK(cudaEventDestroy(stop));
            return ms;
        };

        // ============================================================
        // 4. Warmup + timed runs
        // ============================================================
        int warmups = 5, runs = 20;
        std::cout << "\n[INFO] Starting sequential inference (" << warmups
                  << " warmups, " << runs << " runs)...\n\n";

        for (int w = 0; w < warmups; ++w) {
            nvtxRangePushA(("Warmup " + std::to_string(w + 1)).c_str());
            CUDA_CHECK(cudaMemcpyAsync(dIn0, hIn[0],
                                       volume(inDims[0]) * sizeof(float),
                                       cudaMemcpyHostToDevice, stream));
            for (int i = 0; i < numChunks; ++i) {
                void* src = (i == 0) ? dIn0 : dAct[i - 1];
                runChunk(i, src, dAct[i]);
                CUDA_CHECK(cudaMemcpyAsync(hOut[i], dAct[i],
                                           volume(outDims[i]) * sizeof(float),
                                           cudaMemcpyDeviceToHost, stream));
            }
            CUDA_CHECK(cudaStreamSynchronize(stream));
            nvtxRangePop();
        }

        std::vector<std::vector<float>> chunkTimes(runs);
        std::vector<float> gpuTotals(runs), cpuTotals(runs);

        for (int r = 0; r < runs; ++r) {
            nvtxRangePushA(("Run " + std::to_string(r + 1)).c_str());
            auto t0 = std::chrono::high_resolution_clock::now();

            CUDA_CHECK(cudaMemcpyAsync(dIn0, hIn[0],
                                       volume(inDims[0]) * sizeof(float),
                                       cudaMemcpyHostToDevice, stream));

            std::vector<float> perChunk(numChunks);
            for (int i = 0; i < numChunks; ++i) {
                nvtxRangePushA(("Chunk " + std::to_string(i + 1)).c_str());
                void* src = (i == 0) ? dIn0 : dAct[i - 1];
                perChunk[i] = runChunk(i, src, dAct[i]);
                CUDA_CHECK(cudaMemcpyAsync(hOut[i], dAct[i],
                                           volume(outDims[i]) * sizeof(float),
                                           cudaMemcpyDeviceToHost, stream));
                nvtxRangePop();
            }
            CUDA_CHECK(cudaStreamSynchronize(stream));

            auto t1 = std::chrono::high_resolution_clock::now();
            double cpuMs =
                std::chrono::duration<double, std::milli>(t1 - t0).count();
            double gpuMs = std::accumulate(perChunk.begin(), perChunk.end(), 0.0);
            chunkTimes[r] = perChunk;
            gpuTotals[r] = gpuMs;
            cpuTotals[r] = cpuMs;

            std::cout << "Run " << (r + 1) << ": GPU " << gpuMs
                      << " ms | CPU " << cpuMs << " ms\n";
            nvtxRangePop();
        }

        // ============================================================
        // 5. CSV export
        // ============================================================
        std::ofstream f("chunk_timing.csv");
        f << "run_id,";
        for (int i = 0; i < numChunks; ++i)
            f << "chunk" << (i + 1) << "_gpu_ms"
              << (i + 1 < numChunks ? "," : "");
        f << ",total_gpu_ms,total_cpu_ms\n";

        for (int r = 0; r < runs; ++r) {
            f << (r + 1) << ",";
            for (int i = 0; i < numChunks; ++i)
                f << std::fixed << std::setprecision(4)
                  << chunkTimes[r][i] << (i + 1 < numChunks ? "," : "");
            f << "," << gpuTotals[r] << "," << cpuTotals[r] << "\n";
        }
        std::cout << "📝 Saved detailed timing to chunk_timing.csv\n";

        // cleanup
        for (int i = 0; i < numChunks; ++i) {
            cudaFreeHost(hIn[i]);
            cudaFreeHost(hOut[i]);
            cudaFree(dAct[i]);
        }
        cudaFree(dIn0);
        cudaStreamDestroy(stream);
        std::cout << "✅ Done.\n";
    } catch (std::exception const& e) {
        std::cerr << "Exception: " << e.what() << "\n";
        return 1;
    }
}
